flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/feature/TfIdfTransformer.scala

```diff
@@ -0,0 +1,128 @@
+ * The DataSet will be treated as the corpus of documents. The single words will be filtered against the regex:
+ * <code>
+ * (?u)\b\w\w+\b
+ * </code>
+ * <p>
+ * The TF is the frequence of a word inside one document
+ * <p>
+ * The IDF for a word is calculated: log(total number of documents / documents that contain the word) + 1
+ * <p>
+ * This transformer returns a SparseVector where the index is the hash of the word and the value the tf-idf.
+ * @author Ronny Bräunlich
+ * @author Vassil Dimov
+ * @author Filip Perisic
+ */
+class TfIdfTransformer extends Transformer[(Int, Seq[String]), (Int, SparseVector)] {
+
+  override def transform(input: DataSet[(Int /* docId */ , Seq[String] /*The document */ )], transformParameters: ParameterMap): DataSet[(Int, SparseVector)] = {
+
+    val matchedWordCounts = input.flatMap(t => {
+      t._2.flatMap(s => {
+        "(?u)\\b\\w\\w+\\b".r findAllIn s map (
+          s2 => (t._1, s2.toLowerCase, 1)
+          )
+      })
+    })
+      .filter(t => !transformParameters.apply(StopWordParameter).contains(t._2))
+      //group by document and word
+      .groupBy(0, 1)
+      // calculate the occurrence count of each word in specific document
+      .sum(2)
+
+    val dictionary = matchedWordCounts
+      .map(t => LinkedHashSet(t._2))
+      .reduce((set1, set2) => set1 ++ set2)
+      //to make sure we only create one dictionary
+      .setParallelism(1)
+      .map(set => set.zipWithIndex.toMap)
+      .flatMap(m => m.toList)
+
+    val numberOfWords = matchedWordCounts
+      .map(t => (t._2))
+      .distinct(t => t)
+      .map(t => 1)
+      .reduce(_ + _);
+
+    val idf: DataSet[(String, Double)] = calculateIDF(matchedWordCounts)
+    val tf: DataSet[(Int, String, Int)] = matchedWordCounts
+
+    // docId, word, tfIdf
+    val tfIdf = tf.join(idf).where(1).equalTo(0) {
+      (t1, t2) => (t1._1, t1._2, t1._3.toDouble * t2._2)
+    }
+
+    val res = tfIdf.crossWithTiny(numberOfWords)
+      // docId, word, tfIdf, numberOfWords
+      .map(t => (t._1._1, t._1._2, t._1._3, t._2))
+      //assign every word its position
+      .joinWithTiny(dictionary).where(1).equalTo(0)
+      //join the tuples (docId, word, tfIdf, numberOfWords, index
+      .map(t => (t._1._1, t._1._2, t._1._3, t._1._4, t._2._2))
+      .map(t => {
+      (t._1, List[(Int, Double)]((t._5, t._3)), t._4)
+    })
+      .groupBy(t => t._1)
+      .reduce((t1, t2) => (t1._1, t1._2 ++ t2._2, t1._3))
+      .map(t => (t._1, SparseVector.fromCOO(t._3, t._2.toIterable)))
+
+    res
+  }
+
+  /**
+   * Takes the DataSet of DocId, Word and WordCount and calculates the IDF as tuple of word and idf
+   * @param set
+   */
+  private def calculateIDF(set: DataSet[(Int, String, Int)]) = {
+    val totalNumberOfDocuments = set
+      //map the input only to the docId
+      .map(t => t._1)
+      .groupBy(i => i)
+      //reduce to unique docIds
+      .reduce((t1, t2) => t1)
```

Comment:
Distinct is not possible because the result of the map is not not a tuple type or we don't know how to use it?